Add the constant_closure construction to the Clambda.ustructured_constant type

chambart opened the pull request:
Flambda needs it to be able to explicitly assign a symbol to a constant closure.

This allow for instance to build constants referring to closures:

``` ocaml
let f x = x
let b = (f, f)
```

It also allow to build constant closures with a non empty environment. This can occur in some rare cases in flambda.

This request is following the PR https://github.com/ocaml/ocaml/pull/393, the proposed changes here is the https://github.com/ocaml/ocaml/commit/778eadccc674df6114622b11f2bd967249e01465 patch.

chambart committed:
Make the Uconst_ref constant optionnal.

Cmmgen does not need it, it only helps with unboxing. It is still assumed
to be present inside Closure.

chambart committed:
Add constant closures to Clambda.ustructured_constant

This allow to build constants refering to closures. It also allow to
build constant closures with a non empty environment.

alainfrisch commented:
Can you clarify in which case flambda produces a closure with a non-empty (but constant) environment?  Couldn't one just rewrite the functions to refer to global constants instead of the closure environment?

chambart commented:
It could appear when an inlined version of the function exists referring to the closure using a Project_var that was not simplified. This probably does not appear with the current order of passes, but I would consider that quite fragile to forbid it and I see some potential cases where this might be required in the future.

alainfrisch commented:
> I see some potential cases where this might be required in the future.

Do you agree that one could always (mechanically) rewrite the function to access global constants instead of reading from the closure?  Perhaps this would require some extra processing which is not justified considering the rare gain and the added code complexity, but I want to understand if there is something deeper here or not.

chambart commented:
The function itself won't access the closure. It's the rest of the code that could.

chambart committed:
Continue emiting functions and constants until everything is done

Since constants can contain functions, we need to check if there
is no new functions to handle after emiting constants.

damiendoligez commented:
OK for merging.

smuenzel-js commented:
It looks like the clos_vars argument of Uconst_closure/Const_closure is not used in trunk. Are there still plans to use it or should it be removed?
